app/services/action_builder.rb

```diff
@@ -20,6 +20,15 @@ def build_action(extra_attrs = {})
   def previous_action
```

Comment:
@NealJMD you are right, my comment on memoization is wrong. I meant that the ```if``` statement is always evaluated so there is always a DB hit for _campaign_ (unless it has already been loaded). (It might be better to switch the two conditions, IMHO)

The other part about the loop: when no previous action is found, then ALL campaign pages are looped (here is where my understanding of Champaigns model is incomplete...are there 1,2, 100 pages associated to a campaign?) as there is no _break_ condition.